sdks/java/fn-execution/src/test/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregatorTest.java

```diff
@@ -0,0 +1,385 @@
+                .withOnCompleted(() -> onCompletedWasCalled.set(true))
+                .build());
+
+    // Test that nothing is emitted till the default buffer size is surpassed.
+    FnDataReceiver<byte[]> dataReceiver = registerOutputLocation(aggregator, endpoint, CODER);
+    aggregator.startFlushThread();
+    dataReceiver.accept(new byte[BeamFnDataOutboundAggregator.DEFAULT_BUFFER_LIMIT_BYTES - 50]);
+    assertThat(values, empty());
+
+    // Test that when we cross the buffer, we emit.
+    dataReceiver.accept(new byte[50]);
+    assertEquals(
+        messageWithData(
+            new byte[BeamFnDataOutboundAggregator.DEFAULT_BUFFER_LIMIT_BYTES - 50], new byte[50]),
+        values.get(0));
+
+    // Test that nothing is emitted till the default buffer size is surpassed after a reset
+    dataReceiver.accept(new byte[BeamFnDataOutboundAggregator.DEFAULT_BUFFER_LIMIT_BYTES - 50]);
+    assertEquals(1, values.size());
+
+    // Test that when we cross the buffer, we emit.
+    dataReceiver.accept(new byte[50]);
+    assertEquals(
+        messageWithData(
+            new byte[BeamFnDataOutboundAggregator.DEFAULT_BUFFER_LIMIT_BYTES - 50], new byte[50]),
+        values.get(1));
+
+    // Test that when we close with an empty buffer we only have one end of stream
+    aggregator.sendBufferedDataAndFinishOutboundStreams();
+    assertEquals(endMessage(), values.get(2));
+
+    // Test that we can close twice.
+    aggregator.sendBufferedDataAndFinishOutboundStreams();
+    assertEquals(endMessage(), values.get(2));
+  }
+
+  @Test
+  public void testConfiguredBufferLimit() throws Exception {
+    List<BeamFnApi.Elements> values = new ArrayList<>();
+    AtomicBoolean onCompletedWasCalled = new AtomicBoolean();
+    PipelineOptions options = PipelineOptionsFactory.create();
+    options
+        .as(ExperimentalOptions.class)
+        .setExperiments(Arrays.asList("data_buffer_size_limit=100"));
+    BeamFnDataOutboundAggregator aggregator =
+        new BeamFnDataOutboundAggregator(
+            options,
+            endpoint::getInstructionId,
+            TestStreams.<Elements>withOnNext(values::add)
+                .withOnCompleted(() -> onCompletedWasCalled.set(true))
+                .build());
+    // Test that nothing is emitted till the default buffer size is surpassed.
+    FnDataReceiver<byte[]> dataReceiver = registerOutputLocation(aggregator, endpoint, CODER);
+    aggregator.startFlushThread();
+    dataReceiver.accept(new byte[51]);
+    assertThat(values, empty());
+
+    // Test that when we cross the buffer, we emit.
+    dataReceiver.accept(new byte[49]);
+    assertEquals(messageWithData(new byte[51], new byte[49]), values.get(0));
+    Receiver<?> receiver;
+    if (endpoint.isTimer()) {
+      receiver = Iterables.getOnlyElement(aggregator.outputTimersReceivers.values());
+    } else {
+      receiver = Iterables.getOnlyElement(aggregator.outputDataReceivers.values());
+    }
```

Comment:
Since it is the data receiver why would you need the if?

Reply:
The test is parameterized, and could be with a timers receiver under test.

Reply:
thanks